cohttp-eio/src/server.ml

```diff
@@ -92,22 +92,40 @@ let write output (response : Cohttp.Response.t) body =
   in
   Eio.Buf_write.flush output
 
-let callback { conn_closed; handler } conn input output =
+let callback { conn_closed; handler } ((_, peer_address) as conn) input output =
   let id = (Cohttp.Connection.create () [@ocaml.warning "-3"]) in
   let rec handle () =
     match read input with
-    | `Eof -> conn_closed (conn, id)
+    | `Eof ->
+        let () =
+          Logs.info (fun m ->
+              m "%a: disconnected" Eio.Net.Sockaddr.pp peer_address)
+        in
+        conn_closed (conn, id)
+    | exception Eio.Io (Eio.Net.E (Connection_reset _), _) ->
+        let () =
+          Logs.info (fun m ->
+              m "%a: connection reset" Eio.Net.Sockaddr.pp peer_address)
+        in
+        ()
```

Comment:
When should `conn_closed` be called? I'd assume whenever the connection is closed for any reason, including this and other errors. Possible it should be attached to the switch in `run` instead - then it will always be called when a connection is closed, because the switch is what closes it.

Reply:
I'm unsure, since the argument is not documented in `cohttp/src/server.ml`. I'm not certain it should be called in case the expert response handler finished properly.

Reply:
If we want it called in all cases, I can't rely on the switch AFAICT since `callback` doesn't have it. This is the symptom of a subtler issue I think: when callback finishes - even successfully, it is done buffering the answer but not actually sending it, so it is too early to call `conn_closed`. So unless we `flush` in `callback`, it's tricky to find a relevant point at which to call it.

Looking at the Lwt implementation, it seems to me that we can exit the handling loop in the `Expert` case without calling `conn_closed`.

All in all right now it is my impression that `conn_closed` is not fit for any purpose other than maybe logging the disconnection, but I could be wrong. At the very least it needs a formal specification.

Aside from this, I don't think this PR changes the circumstances in which `conn_closed` is called or not, so I think this can be treated as a separate issue.

Reply:
I'm taking the liberty of resolving this since this PR does not change this behavior: `conn_closed` was only called on proper EOF before, and that didn't change.

I do agree that this is spurious, and `conn_closed` should be properly documented (if not dropped altogether).

Reply:
We can revise the connection closed since it is very limited right now. If my memory is not failing me, the expert mode must be dealt with by hand, while conn_closed was used to do "automatic" cleanup of resources in the least worse way we could think of for the "classical" connections, but we could not really find any good way to properly manage the connections in the lwt case and fully prevent leakages